### ISSUE TITLE
New lint: Repetitive branches

### ISSUE THREAD

**Issue by camsteffen** [rust-lang/rust-clippy#5801]
### What it does

Identifies code where each branch performs the same transformation on some value.

### Categories (optional)

- Kind: clippy::style?

*What is the advantage of the recommended code over the original code*

It is less repetitive and often creates opportunities for further simplification.

### Drawbacks

In some cases, the benefit may be negligible or may differ from stylistic preference.

### Example

```rust
match n {
    1 => Ok("one")
    2 => Ok("two")
    3 => {
        // suppose there is more code here
        Ok("three")
    }
    _ => Ok("other")
}
```

Could be written as:

```rust
let s = match n {
    1 => "one"
    2 => "two"
    3 => {
        // suppose there is more code here
        "three"
    }
    _ => "other"
};
Ok(s)
```

This lint can apply to several constructs: `match`, `if`/`else if`/`else`, `if let`

The transformation could be a wrapper struct or enum or a function call.

The variable value should be a _single_ value. Not, for example, a list of function arguments.

I'm not sure if this should be pedantic. Personally, I would probably want to `#[warn]` it.

**Comment by wiomoc** [comment]
The lint should consider Lifetimes too e.g. : https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=72a2e3511de3b085eea0ddcf9680bf9b
